src/shared/sam/cli/samCliInvoker.ts

```diff
@@ -3,50 +3,43 @@
  * SPDX-License-Identifier: Apache-2.0
  */
 
-import { extensionSettingsPrefix } from '../../constants'
 import { getLogger } from '../../logger'
-import { DefaultSettingsConfiguration } from '../../settingsConfiguration'
 import { ChildProcess, ChildProcessResult } from '../../utilities/childProcess'
-import { DefaultSamCliConfiguration, SamCliConfiguration } from './samCliConfiguration'
 import {
     makeRequiredSamCliProcessInvokeOptions,
     SamCliProcessInvokeOptions,
     SamCliProcessInvoker,
 } from './samCliInvokerUtils'
-import { DefaultSamCliLocationProvider } from './samCliLocator'
 
 import * as nls from 'vscode-nls'
+import { DefaultSamCliConfiguration, SamCliConfiguration } from './samCliConfiguration'
+import { DefaultSettingsConfiguration } from '../../settingsConfiguration'
+import { extensionSettingsPrefix } from '../../constants'
 const localize = nls.loadMessageBundle()
 
-export interface SamCliProcessInvokerContext {
-    cliConfig: SamCliConfiguration
-}
-
-export class DefaultSamCliProcessInvokerContext implements SamCliProcessInvokerContext {
-    public cliConfig: SamCliConfiguration = new DefaultSamCliConfiguration(
-        new DefaultSettingsConfiguration(extensionSettingsPrefix),
-        new DefaultSamCliLocationProvider()
-    )
-}
-
-export function resolveSamCliProcessInvokerContext(
-    params: Partial<SamCliProcessInvokerContext> = {}
-): SamCliProcessInvokerContext {
-    const defaults = new DefaultSamCliProcessInvokerContext()
-
-    return {
-        cliConfig: params.cliConfig || defaults.cliConfig,
-    }
-}
-
 /**
  * Yet another `sam` CLI wrapper.
  *
  * TODO: Merge this with `DefaultSamLocalInvokeCommand`.
  */
 export class DefaultSamCliProcessInvoker implements SamCliProcessInvoker {
     private childProcess?: ChildProcess
-    public constructor(private readonly context: SamCliProcessInvokerContext = resolveSamCliProcessInvokerContext()) {}
+    private readonly context: SamCliConfiguration
+    public constructor(params: {
+        preloadedConfig?: SamCliConfiguration
+        locationProvider?: { getLocation(): Promise<string | undefined> }
+    }) {
+        if (params.preloadedConfig) {
+            this.context = params.preloadedConfig
+        } else if (params.locationProvider) {
+            this.context = new DefaultSamCliConfiguration(
+                new DefaultSettingsConfiguration(extensionSettingsPrefix),
+                params.locationProvider
+            )
+        } else {
+            throw new Error('Invalid constructor args for DefaultSamCliProcessInvoker')
```

Comment:
similar case here: Are `params.preloadedConfig` and `params.locationProvider` mutually exclusive?

Reply:
Would it make more sense to typeguard it? Something like:
```
interface PreloadedConfig {
    preloadedConfig: SamCliConfiguration
}

interface LocationProvider {
    locationProvider: { getLocation(): Promise<string | undefined> }
}

export class DefaultSamCliProcessInvoker implements SamCliProcessInvoker {
    private childProcess?: ChildProcess
    private readonly config: SamCliConfiguration
    public constructor(params: PreloadedConfig | LocationProvider) {
        if (Object.prototype.hasOwnProperty.call(params, 'preloadedConfig')) {
            this.config = (params as PreloadedConfig).preloadedConfig
        } else {
            this.config = new DefaultSamCliConfiguration(
                new DefaultSettingsConfiguration(extensionSettingsPrefix),
                (params as LocationProvider).locationProvider
            )
        }
    }
```

Reply:
That implies that if preloadedConfig is supplied, then locationProvider isn't used. Thus they are mutually exclusive. And if they are, then we just need to throw an error if both are supplied.

Reply:
This is correct. We'd use one or the other.

Reply:
So this and https://github.com/aws/aws-toolkit-vscode/pull/1465#discussion_r558489519 can be resolved just by throwing an error if both are supplied.